### ISSUE TITLE
Persian translation

### ISSUE THREAD

**Issue by mgeisler** [google/comprehensive-rust#671]
Hi @mbpouri, here is an issue for you!

I don't see any translation on https://doc.rust-lang.org/book/appendix-06-translation.html, but perhaps there is a Persian Rust community somewhere where you can find translations of common words :smile: 

The translation is live at https://google.github.io/comprehensive-rust/fa/.

**Comment by mgeisler** [comment]
Let me know if/when you run into https://github.com/rust-lang/mdBook/issues/1486 — then we'll have to work on adding good support for right-to-left languages in mdbook.

**Comment by mbpouri** [comment]
sure, on it. thanks pal @mgeisler

**Comment by mgeisler** [comment]
@DannyRavi, I tried out the translation in #1042 and the text looks like this out of the box:

![image](https://github.com/google/comprehensive-rust/assets/89623/6012220c-9925-404e-a1f8-cbf03d78dd48)

Adding `dir="rtl"` to the `html` tag makes it look like this:

![image](https://github.com/google/comprehensive-rust/assets/89623/d3f41579-4983-4448-a1ce-bacbc263500b)

Does this look more correct to you? I cannot read it, so it's hard for me to judge how it looks :slightly_smiling_face: 

Ideally, we should fix this in the upstream mdbook project (https://github.com/rust-lang/mdBook/issues/1486), but we can perhaps prototype a fix for them here and then contribute it upstream. I'm thinking we can simply use different `index.hbs` template depending on the writing direction.

**Comment by DannyRavi** [comment]
Before starting the translation of the texts. I noticed that the book has a problem with right-to-left texts according to links [#1486](https://github.com/rust-lang/mdBook/issues/1486) and [#671](https://github.com/google/comprehensive-rust/issues/671)
After translating some texts and testing the program on a local server, I realized that the right-to-left texts are not correct and the texts are difficult to understand for Persian speakers. Finally, I had no idea for right-to-left texts.

**Comment by mgeisler** [comment]
Hi @DannyRavi, could you test locally with `mdbook serve` after modifying `theme/index.hbs` like this:

```patch
modified   theme/index.hbs
@@ -1,5 +1,5 @@
 <!DOCTYPE HTML>
-<html lang="{{ language }}" class="sidebar-visible no-js {{ default_theme }}">
+<html lang="{{ language }}" dir="rtl" class="sidebar-visible no-js {{ default_theme }}">
     <head>
         <!-- Book generated using mdBook -->
         <meta charset="UTF-8">
```

This simply adds the `dir="rtl"` attribute to the `html` tag. Is this enough to make the page readable? If so, we could use this as a starting point and fix other issues down the road.

**Comment by mgeisler** [comment]
Hey @DannyRavi, @mbpouri, and @javad-jafari — would you be able to review each other's PRs? We have #1209 (by @DannyRavi) and #1341 (by @javad-jafari) open right now.

**Comment by hamidrezakp** [comment]
Hey there, i just saw this issue and the work you guys all are doing, that mean a lot to Persian community, Thank you all.

Can we use custom font for Persian? it's not pretty good. I'm available to change and fix it if you are okay with it.
I'm suggesting to use `Vazir` fonts family for Persian texts.

**Comment by hamidrezakp** [comment]
@DannyRavi @mbpouri 
Thank you and @javad-jafari for your great work on translation.
I just checked some of translations and it seems and there is a good amount of [ZWNG](https://en.wikipedia.org/wiki/Zero-width_non-joiner) missing in translations, let's fix them before merge to ensure a good quality text.
And I suggest we use `راست` for work ‍`Rust` because reading it is easier and it's pretty common in Persian community right now.

**Comment by mgeisler** [comment]
> Hey there, i just saw this issue and the work you guys all are doing, that mean a lot to Persian community, Thank you all.

Hi @hamidrezakp, thanks for joining in!
 
> Can we use custom font for Persian translation? it's not pretty good. I'm available to change and fix it if you are okay with it. I'm suggesting to use `Vazir` fonts family for Persian texts.

Awesome! If you can write a PR which tweaks the CSS, then I would be happy to add it!

We should ultimately add this upstream in https://github.com/rust-lang/mdBook/ since we're relying on the right-to-left support recently added there (https://github.com/rust-lang/mdBook/pull/1641).

If the fix/improvement is clear, then I suggest adding it upstream directly. Otherwise we can also modify our CSS and theme as needed. Just let me know.

**Comment by mgeisler** [comment]
If we can find 2-3 people who are interested in reviewing, then add yourself to the [CODEOWNERS](https://github.com/google/comprehensive-rust/blob/main/.github/CODEOWNERS) file.

That way you'll be auto-added to future PRs and will thus be able to do everything yourself — @henrif75 and myself should not really be involved in the Persian translation :smile:

**Comment by hamidrezakp** [comment]
> If the fix/improvement is clear, then I suggest adding it upstream directly. Otherwise we can also modify our CSS and theme as needed. Just let me know.

So let me open an issue there and ask them if we can use custom fonts for Persian language; if not, we add it here.

**Comment by hamidrezakp** [comment]
> That way you'll be auto-added to future PRs and will thus be able to do everything yourself — @henrif75 and myself should not really be involved in the Persian translation 😄

Yeah you are right 😄 it's just impossible for you to review or check translations, I think i can find some friends in Persian Rust community that are willing to be maintainer here.

**Comment by mehrad77** [comment]
Hey folks, so existed about this.
I can help with reviewing the translations and other stuff related to `fa` translations.

**Comment by hamidrezakp** [comment]
@mehrad77 I have added your name to reviewers for Persian(#1375 ),
Which one of this PRs do you want to review? 
- #1209 
- #1341

You can assign them to yourself and we will pick others.

**Comment by hamidrezakp** [comment]
Yeah i agree but maybe you can just check last commit (although it contains a lot of non-related things to translation and that's because of poor formatting i guess).

Anyways, That's a good idea to split the translation task into smaller chunks and more organized so contributors can easily pick 'em up and translate.

If you can open a draft PR that mentions what you are working on would be great.

**Comment by DannyRavi** [comment]
> ns.

Hi @hamidrezakp .
Unfortunately, the word  راست  in Farsi, which means  ‍Rust  programming language, is the same as the word   راست  , which means  ‍Right  , and this would increase the error of reading and misunderstand the translated text, and suggestions like  رااست 
have been discussed before and have not reached a conclusion.

**Comment by mehrad77** [comment]
@DannyRavi my suggestion is using `رآست` with the U+0622 instead of U+0627. (As @afshin[ introduced on Mastadon](https://mastodon.social/@afshin@mas.to/110610071106412372)). But I'm see that `رااست` has [more adaption](https://mastodon.social/tags/%D8%B1%D8%A7%D8%A7%D8%B3%D8%AA). I'm in favor of both of these.

**Comment by hamidrezakp** [comment]
@DannyRavi @mehrad77 
I think we should do some polls in communities and social media to see what are everyone okay with.

**Comment by mgeisler** [comment]
> Yeah i agree but maybe you can just check last commit (although it contains a lot of non-related things to translation and that's because of poor formatting i guess).

If the poor formatting is about lines being re-wrapped in strange ways, then we have a solution for that: we recently (#1359) introduced a GitHub check which requires PO files to be formatted consistently. This boils down to following the preparation section of [TRANSLATIONS.md](https://github.com/google/comprehensive-rust/blob/main/TRANSLATIONS.md) and then running

```
dprint fmt
```

Btw, I've found that using [`gh`](https://cli.github.com/) and [`bat`](https://github.com/sharkdp/bat) works very well for reviewing the large diffs:

```
gh pr diff 1341 | bat -l patch
```

is very quick and shows me that changes of #1341 in a nice way.

**Comment by mgeisler** [comment]
> @DannyRavi @mehrad77 I think we should do some polls in communities and social media to see what are everyone okay with.

Great idea! Feel free to use the [discussions here](https://github.com/google/comprehensive-rust/discussions/) or any other convenient place!

**Comment by moaminsharifi** [comment]
First of all, special thanks to @mgeisler for adding dir="rtl|ltr" to the html template

https://github.com/google/comprehensive-rust/issues/671#issuecomment-1661888029

When you add dir="rtl" to html, the page direction will be right to left, but what about the English text (or any other LTR text) between the contents (or vice versa)?

 For example,

 ![Welcome to Comprehensive Rust 🦀 - Comprehensive Rust 🦀 - issue in BiDi](https://github.com/rust-lang/mdBook/assets/11869014/c2c1aeca-21a3-4f29-92ae-1766aa61460a)

In #1413 , I proposed a simple way to handle it with theme but not the most suitable way for doing it.

It's better to mention it here as well.

**Comment by mgeisler** [comment]
Thanks @moaminsharifi for looking at this. As I wrote in #1413, I think we should first understand if the right-to-left support works okayish for fully-translated pages. If it does, then the problem you point out is a temporary problem — if it doesn't, then I would need help from people here to find a good solution.

**Comment by moaminsharifi** [comment]
> Thanks @moaminsharifi for looking at this. As I wrote in #1413, I think we should first understand if the right-to-left support works okayish for fully-translated pages. If it does, then the problem you point out is a temporary problem — if it doesn't, then I would need help from people here to find a good solution.

@mgeisler, For the time being, I think we should close that github issue #1413 until the pages are completely translated, then check it out again.

**Comment by moaminsharifi** [comment]
@mgeisler 

Some of markdown renders to svg with *\`\`\`bob* ,  but I can not find in the translate file.
There is 10 Item like that:
![image](https://github.com/google/comprehensive-rust/assets/11869014/7a93aaa3-884a-4b36-a25d-907ab5624291)
but mostly 4 in each translation file like (`fa.po`):
![in-translation-file](https://github.com/google/comprehensive-rust/assets/11869014/0138be25-472f-43ec-8d5d-5ca3d89de0f5)

Is that something we must add it manually?

**Comment by mgeisler** [comment]
Hi @moaminsharifi, the svgbob images disappeared from the translation pipeline with https://github.com/google/mdbook-i18n-helpers/pull/75 by @dyoo: we now only translate a code block if it contains `"` or `//` since that is a very simple heuristic to recognize code with string literals or comments.

We should restore this and https://github.com/google/mdbook-i18n-helpers/issues/76 is in some sense about this problem: with that feature implemented, we should be able to say "always translate `svgbob` code blocks".

Note that our text extraction tooling will become more fine-grained with https://github.com/google/mdbook-i18n-helpers/pull/109, which will be merged later today or tomorrow after a small rebase.

**Comment by mgeisler** [comment]
[Amazing work](https://github.com/google/comprehensive-rust/pulls?q=is%3Apr+is%3Aclosed+label%3Atranslation%2Ffa) done here by @DannyRavi, @alix1383, @javad-jafari!

When you're ready, please add a link to the [language drop-down](https://github.com/google/comprehensive-rust/blob/main/theme/index.hbs#L178-L203).

**Comment by henrif75** [comment]
> [Amazing work](https://github.com/google/comprehensive-rust/pulls?q=is%3Apr+is%3Aclosed+label%3Atranslation%2Ffa) done here by [@DannyRavi](https://github.com/DannyRavi), [@alix1383](https://github.com/alix1383), [@javad-jafari](https://github.com/javad-jafari)!
> 
> When you're ready, please add a link to the [language drop-down](https://github.com/google/comprehensive-rust/blob/main/theme/index.hbs#L178-L203).

Congratulations! I've created a new issue #2435 and I'll send a PR soon.
